lightning/src/util/persist.rs

```diff
@@ -243,10 +254,873 @@ where
+/// [`OutPoint`]'s transaction ID and index (for example,
+/// `[SOME-64-CHAR-HEX-STRING]_1`).
+///
+/// Each [`ChannelMonitorUpdate`] is stored in a dynamic sub-namespace, as follows:
+///
+///   - namespace: [`CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE`]
+///   - sub-namespace: [the monitor's encoded OutPoint name]
+///
+/// Under that namespace, each update is stored with a left-padded `u64`, like
+/// `00000000000000000021`.
+///
+/// For example, consider this channel:
+///
+///   - Outpoint
+///   - Transaction ID:
+///     `deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f`
+///   - Index: `1`
+///
+/// Full channel monitors would be stored at a single key:
+///
+/// `[CHANNEL_MONITOR_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1`
+///
+/// Updates would be stored as follows (with `/` delimiting
+/// namespace/subnamespace/key):
+///
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000001`
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000002`
+/// `[CHANNEL_MONITOR_UPDATE_PERSISTENCE_NAMESPACE]/deadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33fdeadb33f_1/00000000000000000003`
+/// ... and so on.
+///
+/// # Reading channel state from storage
+///
+/// Channel state is reconstructed by calling
+/// [`Self::read_channel_monitors_with_updates`]. Please see that function's
+/// documentation for more information.
+///
+/// # Pruning stale channel updates
+///
+/// TODO(domz)
+pub struct MonitorUpdatingPersister<K: Deref, L: Deref>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	kv_store: K,
+	logger: L,
+	maximum_pending_updates: u64,
+	last_monitor_update_id: RefCell<HashMap<MonitorName, u64>>,
+}
+
+#[allow(dead_code)]
+impl<K: Deref, L: Deref> MonitorUpdatingPersister<K, L>
+where
+	K::Target: KVStore,
+	L::Target: Logger,
+{
+	/// Reads channel monitors, along with any stored updates for them.
+	///
+	/// # EXTREMELY IMPORTANT
+	/// It is **EXTREMELY IMPORTANT** that your [`KVStore`] implementation uses
+	/// the [`io::ErrorKind::NotFound`] variant correctly: that is,
+	/// when a file is not found, and _only_ in that circumstance (not when
+	/// there is really a permissions error, for example). This is because this
+	/// function does not list updates. Instead, it lists all monitors, and then
+	/// using their stored `update_id`, synthesizes update storage keys, and
+	/// tries them in sequence until one is not found. All _other_ errors will
+	/// be bubbled up in this function's [`io::Result`].
+	pub fn read_channel_monitors_with_updates<
+		ES: Deref + Clone,
+		SP: Deref + Clone,
+		B: Deref,
+		F: Deref + Clone,
+	>(
+		&self,
+		entropy_source: ES,
+		signer_provider: SP,
+		broadcaster: &B,
+		fee_estimator: F,
+	) -> io::Result<
+		Vec<(
```

Comment:
Yeah, it's ugly. `rustfmt` seems like it has some author(s) whose day job pays them by LoC. I just use it because its occasionally obnoxious opinions are outweighed by its convenience, but I am happy to do a manual aesthetic pass as we get close to the finish line?

Reply:
I fixed this (see comment on main thread for more info).